fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCQueryBuilder.java

```diff
@@ -1376,12 +1391,268 @@ private SqlQueryData processMissingParm(Class<?> resourceType, QueryParameter qu
+                        //     ON CLRx.LOGICAL_RESOURCE_ID = CLR<x-1>.LOGICAL_RESOURCE_ID
+                        //     AND
+                        // @formatter:on
+                        whereClauseSegment.append(JOIN).append(previousParm.getModifierResourceTypeName()).append(LOGICAL_RESOURCES)
+                                            .append(AS).append(chainedLogicalResourceVar).append(ON)
+                                            .append(chainedLogicalResourceVar).append(DOT).append(LOGICAL_RESOURCE_ID)
+                                            .append(EQ).append(prevChainedLogicalResourceVar).append(DOT).append(LOGICAL_RESOURCE_ID).append(AND);
+                    }
+                    // Build the rest: CLRx.LOGICAL_ID IN (?)
+                    sqlQueryData = buildChainedIdClause(currentParm, chainedParmVar);
+                }
+
+                if (log.isLoggable(Level.FINE)) {
+                    log.fine("chained sqlQueryData[" + chainedParmVar + "] = " + sqlQueryData.getQueryString());
+                }
+                whereClauseSegment.append(sqlQueryData.getQueryString());
+                bindVariables.addAll(sqlQueryData.getBindVariables());
+           }
+
+            // Insert where clause segment in whole
+            whereClauseSegments.insert(0, whereClauseSegment.toString());
+
+            previousParm = currentParm;
+            parmIndex++;
+        }
+
+        SqlQueryData queryData = new SqlQueryData(whereClauseSegments.insert(0, selectSegments.toString()).toString(), bindVariables);
+        log.exiting(CLASSNAME, METHODNAME, queryData.getQueryString());
+        return queryData;
+    }
+
+    /**
+     * Populates the reference parameter name ID and code system ID sub-segment of the passed where clause segment.
+     *
+     * @param whereClauseSegment - the segment to which the sub-segment will be added
+     * @param queryParmName - the search parameter name
+     * @param resourceTypeName - the resource type of the reference being followed
+     * @param parameterTableAlias - the alias for the parameter table e.g. CPx
+     * @throws FHIRPersistenceException
+     */
+    private void populateReferenceNameAndCodeSystemIdSubSegment(StringBuilder whereClauseSegment, String queryParmName,
+            String resourceTypeName, String parameterTableAlias) throws FHIRPersistenceException {
+        final String METHODNAME = "populateReferenceNameAndCodeSystemIdSubSegment";
+        log.entering(CLASSNAME, METHODNAME, queryParmName);
+
+        Integer parameterNameId = identityCache.getParameterNameId(queryParmName);
+        Integer codeSystemId = getCodeSystemId(resourceTypeName);
+
+        // Build the segment:
+        // CPx.PARAMETER_NAME_ID = <parameter-name-id> AND CPx.CODE_SYSTEM_ID = <code-system_id>
+        whereClauseSegment.append(parameterTableAlias).append(DOT).append(PARAMETER_NAME_ID).append(EQ).append(nullCheck(parameterNameId))
```

Comment:
nullCheck is redundant here. getParameterNameId will never return null.

Reply:
noted